### ISSUE TITLE
Simplified __call__ interface?

### ISSUE THREAD

**Issue by bmcfee** [pescadores/pescador#64]
I've been using pescador a bit over the last few weeks, and find myself wondering if we could tighten up the API a little before the 1.0 release.

Specifically, `streamer.generate()` is a bit verbose; how about adding a `__call__` method so we can instead do something like:

```python
stream = pescador.Streamer(my_generator, my_params...)

for x in stream():
    process(x)
```

This wouldn't help for `cycle` or `tuples`, but it would cut down a bit of boilerplate code for the most common use cases.

**Comment by cjacoby** [comment]
I've also been using it a lot recently, and I like. 👍 

What if we kept the `generator()` interface internally, and did something like this?

```python
def __call__(self, cycle=False, **kwargs):
    if cycle:
        return self.cycle(**kwargs)
    else:
        return self.generate(**kwargs)
```

**Comment by bmcfee** [comment]
I dig it.  Tuples are still an oddball, and I'd prefer to keep that one separate.

How about I implement this after merging the doc updates so that everything can be updated in one pass?

**Comment by cjacoby** [comment]
Roger roger

**Comment by ejhumphrey** [comment]
I realize I slept on this issue the first time around and apologize in advance, but I would be keen to revisit this discussion maybe? 😞  🙏 

Specifically, I'd propose making `core.Streamer` directly iterable, e.g. `[x for x in stream]` is a thing. 

*Commentary* 
- This isn't (necessarily) mutually exclusive with `[x for x in stream(max_batches=10, cycle=True)]` ... though it does violate the "There should be one good way to do a thing" principle. 
- My most common use case is consistent with the default args (which leads me to think that it is _the_ common use case) ... which means everyone using pescador will be writing `for stuff in stream()` all over the place, and those bonus parens smell like a minor anti-pattern. 
- Can the `cycle` and `max_batches` (or `max_samples`, re: #75) go in the init method? I don't immediately see a reason these would change over the lifetime of a Streamer, but am happy to be educated.

**Comment by bmcfee** [comment]
> Can the cycle and max_batches (or max_samples, re: #75) go in the init method? I don't immediately see a reason these would change over the lifetime of a Streamer, but am happy to be educated.

I think this might cause some difficulties with how streamers get used inside `Mux`.  If you're using poisson-distributed sample limits, you would have a different `max_batches` each time the streamer is activated.  Ditto for `cycle`, though it's a bit less obvious there.

Given the above, I'm not sure dropping the call interface (or implementing a direct iterable interface) is the best option, but I'm open to persuasion.

**Comment by ejhumphrey** [comment]
well, here's some simple(r) math: what's the more annoying anti-pattern, a high occurrence of empty parens, or two call signatures for the same functionality?

**Comment by cjacoby** [comment]
The original reason, if I recall correctly, for not allowing you to
__iter__ over streamers directly is language in the Python docs about how
__iter__ should behave, with regards to StopIterations. It was an
intentional design decision based on those Python definitions.

(I'm on my phone or I'd look up the details. I'll get to it in a bit if
@bmcfee doesn't beat me to it)

**Comment by ejhumphrey** [comment]
hrm, well in that case, I would augment my proposal to:
- Iterating on a Streamer behaves as one would expect (it raises StopIteration and runs to exhaustion), and one of...
- *(a)* Explicitly calling the streamer (`__call__`) with args allows the Streamer to augment it's behavior, such as `cycle` itself or terminate early via `max_iter`, which I would assume violates PEP conventions around iterators or *(b)* do we need `__call__` when we have both `generate()` and `cycle()`?

**Comment by cjacoby** [comment]
I would be okay with either solution.

I think (a) is more elegant, and (b) is more explicit.

If I may point back to the original proposal on this Issue, the objective
here is to create or add a simplified API for the End-User which is less
verbose than `streamer.generate()`. I think I would prefer to actually do
both (a) and (b) here, and internally to `Pescador`, or in complicated use
cases, prefer the explicit API, but offer a simplified/concise API as a
"shortcut", which simply calls the Explicit API under the hood.

**Comment by ejhumphrey** [comment]
having thought about it more, I agree with you on 👍  (a) ... a simplified / concise interface for doing non-standard things is nice, and I can imagine that it'd be useful to control these parameters with config-args, rather than augmenting code.

**Comment by ejhumphrey** [comment]
tiny update: I'm having a *really* interesting / useful day catching up to pescador. Given the mix of `Streamer` and generator objects, I would like to emphasize the need for anything that gets referred to as a "stream" to be directly iterable.

**Comment by ejhumphrey** [comment]
closed with #88
